modules/51DegreesRtdProvider.js

```diff
@@ -0,0 +1,255 @@
+const ORTB_DEVICE_TYPE_MAP = new Map([
+  ['Phone', ORTB_DEVICE_TYPE.PHONE],
+  ['Console', ORTB_DEVICE_TYPE.SET_TOP_BOX],
+  ['Desktop', ORTB_DEVICE_TYPE.PERSONAL_COMPUTER],
+  ['EReader', ORTB_DEVICE_TYPE.PERSONAL_COMPUTER],
+  ['IoT', ORTB_DEVICE_TYPE.CONNECTED_DEVICE],
+  ['Kiosk', ORTB_DEVICE_TYPE.OOH_DEVICE],
+  ['MediaHub', ORTB_DEVICE_TYPE.SET_TOP_BOX],
+  ['Mobile', ORTB_DEVICE_TYPE.MOBILE_TABLET],
+  ['Router', ORTB_DEVICE_TYPE.CONNECTED_DEVICE],
+  ['SmallScreen', ORTB_DEVICE_TYPE.CONNECTED_DEVICE],
+  ['SmartPhone', ORTB_DEVICE_TYPE.MOBILE_TABLET],
+  ['SmartSpeaker', ORTB_DEVICE_TYPE.CONNECTED_DEVICE],
+  ['SmartWatch', ORTB_DEVICE_TYPE.CONNECTED_DEVICE],
+  ['Tablet', ORTB_DEVICE_TYPE.TABLET],
+  ['Tv', ORTB_DEVICE_TYPE.CONNECTED_TV],
+  ['Vehicle Display', ORTB_DEVICE_TYPE.PERSONAL_COMPUTER]
+]);
+
+/**
+ * Extracts the parameters for 51Degrees RTD module from the config object passed at instantiation
+ * @param {Object} moduleConfig Configuration object of the 51Degrees RTD module
+ * @param {Object} reqBidsConfigObj Configuration object for the bidders, currently not used
+ */
+export const extractConfig = (moduleConfig, reqBidsConfigObj) => {
+  // Resource key
+  const resourceKey = deepAccess(moduleConfig, 'params.resourceKey');
+  // On-premise JS URL
+  const onPremiseJSUrl = deepAccess(moduleConfig, 'params.onPremiseJSUrl');
+
+  if (!resourceKey && !onPremiseJSUrl) {
+    throw new Error(LOG_PREFIX + ' Missing parameter resourceKey or onPremiseJSUrl in moduleConfig');
+  } else if (resourceKey && onPremiseJSUrl) {
+    throw new Error(LOG_PREFIX + ' Only one of resourceKey or onPremiseJSUrl should be provided in moduleConfig');
+  }
+  if (resourceKey === '<YOUR_RESOURCE_KEY>') {
+    throw new Error(LOG_PREFIX + ' replace <YOUR_RESOURCE_KEY> in configuration with a resource key obtained from https://configure.51degrees.com/tWrhNfY6');
+  }
+
+  return {resourceKey, onPremiseJSUrl};
+}
+
+/**
+ * Gets 51Degrees JS URL
+ * @param {Object} pathData API path data
+ * @param {string} [pathData.resourceKey] Resource key
+ * @param {string} [pathData.onPremiseJSUrl] On-premise JS URL
+ * @returns {string} 51Degrees JS URL
+ */
+export const get51DegreesJSURL = (pathData) => {
+  if (pathData.onPremiseJSUrl) {
+    return pathData.onPremiseJSUrl;
+  }
+  return `https://cloud.51degrees.com/api/v4/${pathData.resourceKey}.js`;
+}
+
+/**
+ * Injects 51Degrees script into the document head
+ * @param {string} url 51Degrees JS URL
+ * @returns {Promise} Promise that resolves when the script is loaded
+ */
+export const inject51DegreesScript = (url) => {
+  const script = document.createElement('script');
+  script.src = url;
+  script.async = true;
+  document.head.appendChild(script);
+
+  return new Promise((resolve, reject) => {
+    script.onload = resolve;
+    script.onerror = reject;
+  });
+}
+
+/**
+ * Check if meta[http-equiv="Delegate-CH"] tag is present in the document head and points to 51Degrees cloud
```

Comment:
What's going on here, why not just get the high entropy hints in js

Reply:
Thanks for the comment.  Agreed that in general it's a good idea to use `getHighEntropyValues` API.  The injected script actually uses it as a fallback in the case when navigator did not send User-Agent Client Hints as part of the first request.  Having the navigator send the User-Agent Client Hints by itself seems to be a bit more performant and also transparent to the user that they give permission to a 3rd party to process those as prescribed by the [spec](https://wicg.github.io/client-hints-infrastructure/#delegate-ch-algo)

Reply:
FWIW - PBJS-core will add high entropy device.sua if the publisher sets some config...

```javascript
pbjs.setConfig({
       firstPartyData: {
         uaHints: ['platform', 'architecture', 'model', 'platformVersion', 'fullVersionList']
       }
});
```

This may be something you want to suggest in your doc?

Reply:
Thanks for raising this, @bretg. A lengthy comment ahead, but we hope it will become a basis for more documentation. Breaking it down below. 

## PBS modules

The [First Party Data Enrichment](https://docs.prebid.org/dev-docs/modules/enrichmentFpdModule.html) module with uaHints mentioned above is very relevant for 51D Prebid Server Modules, and we'll actually add it to the server modules docs, because 51D PBS modules are more precise when `device.sua` is present in the request.

## Prebid.js module

51D Prebid.js module injects a dynamically generated script based on the evidence available. It also falls back to the GetHighEntropyValues API, but does not rely on it as a first (or only) choice route - please see the illustrative cases below.  Albeit it seems easier, GHEV API is not supported by all browsers (so the decision to call it should be conditional) and also even in Chrome this API will likely be a subject to the Privacy Budget in the future.

In summary we use `Delegate-CH` http-equiv as the preferred method of obtaining the necessary evidence because it is the fastest method, and future proof. `Delegate-CH` enables the Chromium-based browser to send the User-Agent Client Hints to a 3rd party right away on the very first request unlike `Accept-CH` + `Permissions-Policy` http response headers which would let it send them on subsequent requests. Hence this check in the code that outputs the warning in debug mode. 

## Illustrative Cases

* if the device is iPhone/iPad then there is no point checking for or calling GetHighEntropyValues at the moment because iOS does not support this API. However this might change in the future.  
Platforms like iOS require additional techniques to identify the model which are not covered via a single API call, and change from version to version of the operating system and browser rendering engine. **When used with iOS 51Degrees resolves the [iPhone/iPad model groups](https://51degrees.com/documentation/4.4/_device_detection__features__apple_device_table.html) using these techniques.** That is one of the benefits the module brings to the Prebid community as most solutions do not resolve iPhone/iPad model groups. More on Apple Device Detection [here](https://51degrees.com/documentation/4.4/_device_detection__features__apple_detection.html).

* if the browser is Brave on Android then there is similarly no point requesting GHEV as the API is not supported.

* if the browser is Chrome then the `Delegate-CH` if enabled by the publisher would enable the browser to provide the necessary evidence. However if this is not implemented - then the dynamic script would fall back to GHEV which is slower.

As the code for these features changes frequently the relevant snippets are in the data file rather than the core source code. As the web diverges the ability to dynamically adapt is going to be very important. Reducing the depency on Javascript and maximising server side stable controlled environments is part of the 51Degrees approach.

Reply:
@jwrosewell in general it is bad form to mark reviewer comments as resolved. We let reviewers decide if their comments are resolved.

Reply:
Unresolving to make easier to find for future readers.